crates/api_common/src/request.rs

```diff
@@ -53,43 +54,53 @@ pub async fn fetch_link_metadata(url: &Url, context: &LemmyContext) -> LemmyResu
   // We only fetch the first 64kB of data in order to not waste bandwidth especially for large
   // binary files
   let bytes_to_fetch = 64 * 1024;
-  let response = context
-    .client()
-    .get(url.as_str())
-    // we only need the first chunk of data. Note that we do not check for Accept-Range so the
-    // server may ignore this and still respond with the full response
-    .header(RANGE, format!("bytes=0-{}", bytes_to_fetch - 1)) /* -1 because inclusive */
-    .send()
-    .await?
-    .error_for_status()?;
+  let mut response = Some(
+    context
+      .client()
+      .get(url.as_str())
+      // we only need the first chunk of data. Note that we do not check for Accept-Range so the
+      // server may ignore this and still respond with the full response
+      .header(RANGE, format!("bytes=0-{}", bytes_to_fetch - 1)) /* -1 because inclusive */
+      .send()
+      .await?
+      .error_for_status()?,
+  );
 
-  // In some cases servers send a wrong mime type for images, which prevents thumbnail
-  // generation. To avoid this we also try to guess the mime type from file extension.
-  let content_type = mime_guess::from_path(url.path())
-    .first()
-    // If you can guess that its an image type, then return that first.
-    .filter(|guess| guess.type_() == mime::IMAGE)
-    // Otherwise, get the content type from the headers
-    .or(
-      response
-        .headers()
+  let content_type: Option<Mime> = {
+    let mut mime = response.as_ref().and_then(|m| {
+      m.headers()
         .get(CONTENT_TYPE)
         .and_then(|h| h.to_str().ok())
-        .and_then(|h| h.parse().ok()),
-    );
+        .and_then(|h| h.parse().ok())
+    });
+
+    // If a server is serving `application/octet-stream`, it's likely a mistake,
+    // so we try to guess the file type from its magic number.
+    if mime
+      .as_ref()
+      .is_some_and(|m: &Mime| m.subtype() == "octet-stream")
+    {
+      // Don't need to fetch as much data for this as we do with opengraph
+      let octet_bytes =
+        collect_bytes_until_limit(response.take().ok_or(FederationError::Unreachable)?, 512)
+          .await?;
+      mime = infer::get(&octet_bytes).map_or(mime, |t| t.mime_type().parse().ok());
+    }
+    mime
+  };
 
   let opengraph_data = {
     // if the content type is not text/html, we don't need to parse it
     let is_html = content_type
       .as_ref()
       .map(|c| {
         (c.type_() == mime::TEXT && c.subtype() == mime::HTML)
-      ||
-      // application/xhtml+xml is a subset of HTML
-      (c.type_() == mime::APPLICATION && c.subtype() == "xhtml")
+	  ||
+	// application/xhtml+xml is a subset of HTML
+	  (c.type_() == mime::APPLICATION && c.subtype() == "xhtml")
       })
       .unwrap_or(false);
-    if !is_html {
+    if !is_html || response.is_none() {
```

Comment:
`response` is hardcoded to `Some`

Reply:
The conditional call to [take](https://doc.rust-lang.org/std/option/enum.Option.html#method.take)  further up changes it a None.

Reply:
You shouldn't need take, as there should always be a response. You just might need to copy/clone the response, or even better use its reference if possible.